src/Azure/Queues/AzureServiceBusQueue.cs

```diff
@@ -7,73 +7,89 @@
+using Nito.AsyncEx;
 
 namespace Foundatio.Queues {
     public class AzureServiceBusQueue<T> : QueueBase<T> where T : class {
         private readonly string _queueName;
+        private readonly string _connectionString;
         private readonly NamespaceManager _namespaceManager;
-        private readonly QueueClient _queueClient;
-        private QueueDescription _queueDescription;
+        private QueueClient _queueClient;
         private long _enqueuedCount;
         private long _dequeuedCount;
         private long _completedCount;
         private long _abandonedCount;
         private long _workerErrorCount;
         private readonly int _retries;
+        private readonly RetryPolicy _retryPolicy;
         private readonly TimeSpan _workItemTimeout = TimeSpan.FromMinutes(5);
+        private readonly AsyncLock _lock = new AsyncLock();
 
-        public AzureServiceBusQueue(string connectionString, string queueName = null, int retries = 2, TimeSpan? workItemTimeout = null, bool shouldRecreate = false, RetryPolicy retryPolicy = null, ISerializer serializer = null, IEnumerable<IQueueBehavior<T>> behaviors = null, ILoggerFactory loggerFactory = null) : base(serializer, behaviors, loggerFactory) {
+        public AzureServiceBusQueue(string connectionString, string queueName = null, int retries = 2, TimeSpan? workItemTimeout = null, RetryPolicy retryPolicy = null, ISerializer serializer = null, IEnumerable<IQueueBehavior<T>> behaviors = null, ILoggerFactory loggerFactory = null) : base(serializer, behaviors, loggerFactory) {
+            _connectionString = connectionString;
             _queueName = queueName ?? typeof(T).Name;
             _namespaceManager = NamespaceManager.CreateFromConnectionString(connectionString);
             _retries = retries;
-            if (workItemTimeout.HasValue && workItemTimeout.Value < TimeSpan.FromMinutes(5))
+            _retryPolicy = retryPolicy;
+
+            if (workItemTimeout.HasValue && workItemTimeout.Value < TimeSpan.FromMinutes(5)) {
                 _workItemTimeout = workItemTimeout.Value;
+            }
+        }
 
-            if (_namespaceManager.QueueExists(_queueName) && shouldRecreate)
-                _namespaceManager.DeleteQueue(_queueName);
+        protected override async Task EnsureQueueCreatedAsync(CancellationToken cancellationToken = new CancellationToken()) {
+            if (_queueClient != null) {
+                return;
+            }
 
-            if (!_namespaceManager.QueueExists(_queueName)) {
-                _queueDescription = new QueueDescription(_queueName) {
-                    MaxDeliveryCount = retries + 1,
-                    LockDuration = _workItemTimeout
-                };
-                _namespaceManager.CreateQueue(_queueDescription);
-            } else {
-                _queueDescription = _namespaceManager.GetQueue(_queueName);
+            using (await _lock.LockAsync(cancellationToken)) {
+                if (_queueClient != null) {
```

Comment:
This is good, I had a question about this while reviewing the code and wondering what would happen in a multi threaded environment 👍